src/Build/Definition/Project.cs

```diff
@@ -1117,33 +1129,169 @@ public List<GlobResult> GetAllGlobs(string itemType)
+
+            public IEnumerable<IMSBuildGlob> Globs => _globs.ToImmutable();
+            public IEnumerable<string> FragmentStrings => _fragmentStrings.ToImmutable();
+
+            public static CumulativeRemoveElementData New()
+            {
+                return new CumulativeRemoveElementData
+                {
+                    _globs = ImmutableList.CreateBuilder<IMSBuildGlob>(),
+                    _fragmentStrings = ImmutableHashSet.CreateBuilder<string>()
+                };
+            }
+
+            public void CumulateInformationFromRemoveItemSpec(EvaluationItemSpec removeSpec)
+            {
+                var removeSpecFragmentStrings = removeSpec.FlattenFragmentsAsStrings();
+                var removeGlob = removeSpec.ToMSBuildGlob();
+
+                _globs.Add(removeGlob);
+
+                foreach (var removeFragment in removeSpecFragmentStrings)
+                {
+                    _fragmentStrings.Add(removeFragment);
+                }
+            }
+        }
+
         private List<GlobResult> GetAllGlobs(List<ProjectItemElement> projectItemElements)
         {
-            return projectItemElements
-                .AsParallel()
-                .AsOrdered()
-                .SelectMany(GetAllGlobs)
-                .ToList();
+            if (projectItemElements.Count == 0)
+            {
+                return new List<GlobResult>();
+            }
+
+            // Scan the project elements in reverse order and build globbing information for each include element.
+            // Based on the fact that relevant removes for a particular include element (element A) consist of:
+            // - all the removes seen by the next include statement of A's type (element B which appears after A in file order)
+            // - new removes between A and B (removes that apply to A but not to B. Spacially, these are placed between A's element and B's element)
+
+            // Example:
+            // 1. <A Include="A"/>
+            // 2. <A Remove="..."/> // this remove applies to the A includes
+            // 3. <A Include="B"/>
+            // 4. <A Remove="..."/> // this remove applies to the A and B includes
+            // 5. <A Include="C"/>
+            // 6. <A Remove="..."/> // this remove applies to the A, B, and C includes
+            // So A's applicable removes are composed of:
+            // 
+            // The applicable removes for the element at position 1. are composed of:
+            // - all the removes seen by the next include statement of A's type (element B which appears after A in file order). In this example that's Removes 4 and 6.
+            // - new removes between A and B. In this example that's Remove 2.
+
+            // use immutable lists because there will be a lot of structural sharing between includes which share increasing subsets of corresponding remove elements
+            // item type -> aggregated information about all removes seen so far for that item type
```

Comment:
I'd keep it as a comment on the reason why immutable types are really good here.